### ISSUE TITLE
Default values are not set.

### ISSUE THREAD

**Issue by relief-melone** [thesayyn/protoc-gen-ts#107]
Alot of implementations omit default values when serializing protobufs. So e.g. enums set to 0 will not be sent over the wire. When the message is deserialized this missing value should be set to the default value of 0. However default values will not be set by protoc-gen-ts. This causes problems with interoperability between e.g. C# and JavaScript/TypeScript as in our case. I investigated the generated code a bit and my suspect is this

**src/compiler/descriptor.js**
```js
// ...
 constructor(data?: any[] | {
            my_enum?:  Payload.SomeEnum
        }) {
            super();
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], []);
            if (!Array.isArray(data) && typeof data == "object") {
                if ("my_enum" in data && data.my_enum != undefined) {
                    this.my_enum = data.my_enum;
                }
            }
        }

// ...
```

If I read this correct then the value is only set if it is there but according to the specification it should be set to the default value 0.

**Comment by thesayyn** [comment]
Hmm. this looks like has something to do with generated getters. for proto3 we are supposed to call `getFieldWithDefault` instead of `getField`.  protoc supposed to set default_value for proto3 but seems like it doesn’t. 

here: https://github.com/thesayyn/protoc-gen-ts/blob/b47d74afc5b35cf13bd66a57b2f77c56debee212/src/descriptor.js#L889

would like to fix it and send a PR with tests included?

**Comment by thesayyn** [comment]
I'd like to have presence getters in here if you are willing to send a separate PR for it.
See: https://github.com/thesayyn/protoc-gen-ts/issues/80#issuecomment-1021039435

**Comment by thesayyn** [comment]
> Yea but then there is a problem when you can't distinguish between not set and default value

this is exactly what getFieldWithDefault does. If the field is present it gives you the value of it but if not it will return you whatever you passed to it as the third argument.

https://github.com/protocolbuffers/protobuf/blob/2f91da585e96a7efe43505f714f03c7716a94ecb/js/message.js#L873-L880

**Comment by thesayyn** [comment]
> Also I found that proto3 js uses functions like `setProto3StringField` and `setProto3IntField` etc., but I have no idea if it's necessary.

they are basically guarding against serializing default values. it is a small optimization that I didn't bother implementing.

**Comment by Yurzel** [comment]
@thesayyn is it possible to contact you on any other platform? I'm having issues with compilation. Otherwise AST should be done.

**Comment by thesayyn** [comment]
grpc has no slack channel. I am on bazel slack channel. you can reach me there.
http://slack.bazel.build/

**Comment by thesayyn** [comment]
in order to close this issue, protoc-gen-ts should be fixed to not serialize default values and should return defaults when field has no presence

**Comment by thesayyn** [comment]
but as far as I am concerned, there shouldn't be any interoperability issues between protoc-gen-ts and other language implementations.

**Comment by thesayyn** [comment]
```patch
diff --git a/src/descriptor.ts b/src/descriptor.ts
index f1cb6da..2d1c68f 100644
--- a/src/descriptor.ts
+++ b/src/descriptor.ts
@@ -916,7 +916,13 @@ function createGetterCall(
       ts.factory.createNumericLiteral(fieldDescriptor.number),
     ];
 
-    if (fieldDescriptor.default_value) {
+    let default_val = fieldDescriptor.default_value
+
+    if (!default_val) {
+        // assign default values
+    }
+
+    if (default_val) {
       getterMethod = "getFieldWithDefault";
       let _default: ts.Expression;
 

```

**Comment by TMQPH** [comment]
@thesayyn Will you add this to the next release?

**Comment by atakurt034** [comment]
@thesayyn Will you be adding this to the next release?

**Comment by relief-melone** [comment]
sorry that there has been so little activity on this one from me. I'd love to help out as I really like this project. Unfortunately I'm so packed with work that I don't find the time to get into to topic enough to contribute (my biggest issue is not beeing familiar with the build and test frameworks). I hope I'll find some more time soon.

Additionaly I just noticed that boolean values seem to have the same problem. false seems to dissapear as its beeing the default value. I can open up a separate issue for that just wanted to clarifiy hiere first

**Comment by ciricc** [comment]
+1

**Comment by thesayyn** [comment]
fixed at #146 and released in 0.8.5
